app/services/saml_request_validator.rb

```diff
@@ -0,0 +1,41 @@
+class SamlRequestValidator
+  include ActiveModel::Model
+
+  validate :authorized_service_provider
+  validate :authorized_authn_context
+
+  def call(service_provider:, authn_context:)
+    self.service_provider = service_provider
+    self.authn_context = authn_context
+
+    @success = valid?
+
+    result
+  end
+
+  private
+
+  attr_accessor :service_provider, :authn_context
+  attr_reader :success
+
+  def result
+    {
+      authn_context: authn_context,
+      errors: errors.messages.values.flatten,
+      service_provider: service_provider.issuer,
+      valid: success
+    }
+  end
+
+  def authorized_service_provider
+    return if service_provider.valid?
+
+    errors.add(:service_provider, 'Unauthorized Service Provider')
```

Comment:
This error is not shown to users. It's only for us to track this event in our analytics. What the user sees is a blank page due to `render nothing: true`. Obviously that's not a great UX, but we haven't yet decided what we should do in this scenario (when the service provider and/or authentication context are invalid, which would be an error caused by the Service Provider sending an invalid SAML request, not the user). See issue 912 if you have any thoughts.